### DISCUSSION TITLE
How to use packges from DefintelyTyped

### DISCUSSION THREAD

**Discussion by glenngijsberts** [DefinitelyTyped/DefinitelyTyped#64252] [_Misc]
I'm wondering how I should use the type packages from DefintelyTyped. I noticed that almost all DT packages that I use (`facebook-js-sdk`, `google.accounts`, `intercom-web`) are using namespaces. When installing, these packages give me more details on the types when using it from the window, but those packages are not exporting any types, so I can't type a function that's using one of the values that are declared in the namespace for example. Also, since I'm loading in the scripts with script tag (and there is no NPM package available for this) I don't have access to types from the scripts themselves.

How should I approach a situation like that? Can I somehow use the namespace to get types? Do I need to write my own types? Let me know please!

**Comment by bershanskiy** [comment]
> Also, since I'm loading in the scripts with script tag (and there is no NPM package available for this) I don't have access to types from the scripts themselves.

I'm not sure what you mean by this. TypeScript type annotations are removed when TypeScript files are compiled to JavaScript.

> How should I approach a situation like that? Can I somehow use the namespace to get types? Do I need to write my own types? Let me know please!

You should not have to explicitly import types. Your TypeScript config already should have them (usually in `tsconfg.json` file). Then you can just refer to the types directly. For example, look at this:
`tsconfig.json` imports `@types/chrome` by declaring `types` attribute with string `chrome`.
https://github.com/darkreader/darkreader/blob/main/src/tsconfig.json
Then the actual code uses the types `chrome.*` directly as if they were imported:
https://github.com/darkreader/darkreader/blob/main/src/background/tab-manager.ts#L204

**Reply by minicatsCB** [reply]
I still don't understand well the difference between using modules and namespaces in DT. In my case I'm using `@types/google.accounts` and checking the `.d.ts` file I could see no `export` keyword inside it. Maybe it's me just expecting to have a type somehow like `GoogleAccounts` to let know TS that, when the corresponding GSI client finishes loading, there will be a global variable available named `google` of type `GoogleAccounts`. In code:

```
declare var google: GoogleAccounts;
// ..
google.accounts.id.initialize(..);
```
And even each of those properties having their own type.

Is it correct to expect this from DT?

**Reply by bershanskiy** [reply]
> there will be a global variable available named google of type GoogleAccounts
> Is it correct to expect this from DT?

TypeScript types are removed from code during compilation (transpilation) to JavaScript. `declare` is used to inform (declare to) TypeScript compiler that there will be a variable/class/function with a specific name and value, without actually creating it. That variable/class/function somehow needs to be present in the context already (e.g., if page has two `<script>` tags and one is a library which creates a bunch of global functions, then the other script can just `declare` existence of these functions).

**Reply by minicatsCB** [reply]
Thank you, that is clear for me now.

How about the difference between using types and namespaces in DT? I still don't understand why I can't do something like `declare var google: GoogleAccounts;` (note the "GoogleAccounts" type) and instead must do `declare var google: any;`. I now that with transpilation this is not a concern, but as a developer, having specific types in code makes it more intuitive at first sight. The generic `any` type doesn't help too much on that...
